spatialdata/_core/aggregate.py

```diff
@@ -0,0 +1,164 @@
+
+        For points, this could be probe intensity.
+    agg_func
+        Aggregation function to apply over point values, e.g. "mean", "sum", "count".
+        Passed to pandas.DataFrame.groupby.agg.
+
+    Returns
+    -------
+    AnnData of shape (by.shape[0], values[id_key].nunique())])
+    """
+    # TODO: Check that values are in the same space
+    # Dispatch
+    by_type = get_schema(by)
+    values_type = get_schema(values)
+    if by_type is ShapesModel:
+        if values_type is PointsModel:
+            return _aggregate_points_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        elif values_type is ShapesModel:
+            return _aggregate_shapes_by_shapes(values, by, id_key, value_key=value_key, agg_func=agg_func)
+        else:
+            raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+    else:
+        raise NotImplementedError(f"Cannot aggregate {values_type} by {by_type}")
+
+
+def _aggregate_points_by_shapes(
+    points: ddf.DataFrame | pd.DataFrame,
+    shapes: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    # Have to get dims on dask dataframe, can't get from pandas
+    dims = sd.get_dims(points)
+    if isinstance(points, ddf.DataFrame):
+        points = points.compute()
+    points = gpd.GeoDataFrame(points, geometry=gpd.points_from_xy(*[points[dim] for dim in dims]))
+    return _aggregate(points, shapes, id_key, value_key, agg_func)
+
+
+def _aggregate_shapes_by_shapes(
+    values: gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    *,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    def circles_to_polygons(df: gpd.GeoDataFrame) -> gpd.GeoDataFrame:
+        # We should only be buffering points, not polygons. Unfortunately this is an expensive check.
+        values_geotypes = list(values.geom_type.unique())
+        if values_geotypes == ["Point"]:
+            df = df.set_geometry(df.geometry.buffer(df[ShapesModel.RADIUS_KEY]))
+        elif "Point" in values_geotypes:
+            raise TypeError("Geometry contained shapes and polygons.")
+        return df
+
+    values = circles_to_polygons(values)
+    by = circles_to_polygons(by)
+
+    return _aggregate(values, by, id_key, value_key, agg_func)
+
+
+def _aggregate(
+    value: gpd.GeoDataFrame,
+    by: gpd.GeoDataFrame,
+    id_key: str,
+    value_key: str | None = None,
+    agg_func: str = "count",
+) -> ad.AnnData:
+    """
+    Inner function to aggregate geopandas objects.
+
+    See docstring for `aggregate` for semantics.
+
+    Parameters
+    ----------
+    value
+        Geopandas dataframe to be aggregated. Must have a geometry column.
+    by
+        Geopandas dataframe to group values by. Must have a geometry column.
+    id_key
+        Column in value dataframe to group values by.
```

Comment:
Do you mean "column in by dataframe"?

Reply:
No, this is from `values`. This is so you can do things like count transcripts (`values`) by gene (`values[id_key]`) by cell (`by`).

Reply:
Ah got it! In the case of points in the model we have a `FEATURE_KEY` that tells which is the preferred aggregation column (see discussion here https://github.com/scverse/spatialdata/pull/113#issuecomment-1419502673), so I would put `id_key` optional and use that by default in the case of points, if available.

Reply:
I think it's right that there are probably good implicit defaults here. I think this get's more complicated for other cases.

In essence, `id_key` is needed when we have a one-dimensional value to aggregate. But there cases where we have two dimensional values to aggregate. 

* Where `values` is an image, the values are channels so we'd likely the output to be (observations in by, channels in values) (we also have no need for `values_key` here)
* Where `values` is a Shapes or Labels SpatialData. Here, we could be aggregating `X`.

I think the values case can wait for now, but these are good things to think about. Especially for shapes how we aggregate values other than `X`.